analyzer/evmnfts/httpmisc/http_misc.go

```diff
@@ -0,0 +1,49 @@
+package httpmisc
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"time"
+)
+
+const ClientTimeout = 30 * time.Second
+
+func GetWithContextWithClient(ctx context.Context, client *http.Client, url string) (*http.Response, error) {
+	req, err := http.NewRequestWithContext(ctx, "GET", url, nil)
+	if err != nil {
+		return nil, err
+	}
+	return client.Do(req)
+}
+
+type NotOKError struct {
+	// Note: .error is the implementation of .Error, .Unwrap etc. It is not
+	// in the Unwrap chain. Use something like
+	// `NotOKError{fmt.Errorf("...: %w", err)}` to set up an
+	// instance with `err` in the Unwrap chain.
+	error
+}
+
+func (err NotOKError) Is(target error) bool {
+	if _, ok := target.(NotOKError); ok {
+		return true
+	}
+	return false
+}
+
+func ResponseOK(resp *http.Response) error {
+	if resp.StatusCode >= 500 {
+		if err := resp.Body.Close(); err != nil {
+			return fmt.Errorf("HTTP closing body: %w", err)
+		}
+		return fmt.Errorf("HTTP %d", resp.StatusCode)
+	}
+	if resp.StatusCode != 200 {
+		if err := resp.Body.Close(); err != nil {
+			return fmt.Errorf("HTTP closing body: %w", err)
+		}
+		return NotOKError{fmt.Errorf("HTTP %d", resp.StatusCode)}
```

Comment:
What is the point of NotOKError? It's an "enum" with a single value, it's produced seemingly inconsistently (I don't understand why the ability to read the response body of e.g. a HTTP 404 would differentiate between a `NotOKError` and a `fmt.Errorf`) and nobody "reads" it, i.e. checks for its presence.
Can we get rid of it and reintroduce when/if requirements are clearer?

Reply:
meant for a future update to separate whether to retry. statuses 5xx indicate server error, which would be a good candidate for retry.

4xx indicates the server isn't satisfied with our request, so it's probably better to give up. Come to think of it, 429 might be better in the other category.

3xx is supposed to be a redirect, but the HTTP client ought to handle that for us. if for whatever reason it didn't, then it's not like we can handle it out here, so it's probably better to give up.

2xx other than 200 ("OK") indicates some kind of successful request but where it doesn't actually give us a file. we're always looking for a file, so it's probably better to give up.

1xx is some message used in the HTTP protocol but aren't formally the response. I suspect the HTTP client handles these internally, but if it somehow gets out here, then we can't handle it, so it's probably better to give up.

These "give up" cases are lumped into a `NotOKError`

Reply:
Alright thanks for the explanation. (And great catch with `HTTP 429 Too Many Requests`! I agree. Wanna encode that now?)
In that case, two small followups:
- Let's rename `NotOKError` please. The name misled me when reading because I think of "error" and "not ok" as pretty synonymous. Maybe `NonRetriableError`? Or we could reuse the name from the RPC cache: `DeterministicError`. Or similar.
- In the non-5xx case, even if we fail to `Close()` the body, I'd say that's still a `DeterministicError`. The HTTP status code is all we need to determine (heh) that. We can just silently log the `Close()` problem maybe? Or even ignore it altogether, assuming some resource cleanup happens behind the scenes.

Reply:
changed to ResourceError, to convey that the server successfully answered about the Resource that the URL Locates, but we still find it unsatisfactory